### PR TITLE
Lab 03 - Letter frequency

### PR DESCRIPTION
**Pull request by JackLeeHal**
Fixes #63

#### Changes proposed in this PR:

* Implement the letter frequency
* Add test cases
* Fix CI issue

### CODE REVIEW

File: 03_letters/jacklee/letter.go
Comment (anzboi) [inline 03_letters/jacklee/letter.go:27]: I was going to ask you to change this to use a fixed size array, populated directly with an index variable, then did some research and found out this method isn't much slower (benchmarks ~20% slower, pretty sure no resizing is going on). Nice work on finding this method 👍. I'll leave it up to you whether you want to change this to gain that 20% speedup.

**Review by anzboi — APPROVED** [review]
LGTM 👍
See the comment about your array populating method. It's a little slower than optimal but not a big issue.

File: 03_letters/jacklee/letter.go
Comment (anzdaddy) [inline 03_letters/jacklee/letter.go:27]: Why not `[]rune`?

File: 03_letters/jacklee/letter_test.go
Comment (anzdaddy) [inline 03_letters/jacklee/letter_test.go:42]: No need to explicitly say what went wrong ("unexpected output") since `require.Equal` provides a detailed description and the maintainer can read the code. Instead, include any necessary context in the error output to make it easy for the engineer to identify where the problem occurred:
```go
for i, d … {
	⋮
	require.Equal(t, d.letter, l, "%d", i)
```

File: 03_letters/jacklee/letter_test.go
Comment (anzdaddy) [inline 03_letters/jacklee/letter_test.go:52]: No need to explicitly say what went wrong ("unexpected output") since `require.Equal` provides a detailed description and the maintainer can read the code. Instead, include any necessary context in the error output to make it easy for the engineer to identify where the problem occurred:
```go
for i, d … {
	⋮
	require.Equal(t, d.letter, l, "%d", i)
```

**Review by anzdaddy — CHANGES_REQUESTED** [review]

**Comment by juliaogris** [comment]
Sorry, I will close this PR now in preparation for the next training in a few weeks in Melbourne - feel free to re-open it if you want to continue work on it.

**Pull request closed without merging**
